ad/internal/winrmhelper/winrm_group_membership.go

```diff
@@ -0,0 +1,197 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/masterzen/winrm"
+)
+
+type GroupMembership struct {
+	GroupGUID    string
+	GroupMembers []*GroupMember
+}
+
+type GroupMember struct {
+	SamAccountName string `json:"SamAccountName"`
+	DN             string `json:"DistinguishedName"`
+	GUID           string `json:"ObjectGUID"`
+	Name           string `json:"Name"`
+}
+
+func groupExistsInList(g *GroupMember, memberList []*GroupMember) bool {
+	for _, item := range memberList {
+		if g.GUID == item.GUID {
+			return true
+		}
+	}
+	return false
+}
+
+func diffGroupMemberLists(expectedMembers, existingMembers []*GroupMember) ([]*GroupMember, []*GroupMember) {
+	var toAdd, toRemove []*GroupMember
+	for _, member := range expectedMembers {
+		if !groupExistsInList(member, existingMembers) {
+			toAdd = append(toAdd, member)
+		}
+	}
+
+	for _, member := range existingMembers {
+		if !groupExistsInList(member, expectedMembers) {
+			toRemove = append(toRemove, member)
+		}
+	}
+
+	return toAdd, toRemove
+}
+
+func unmarshalGroupMembership(input []byte) ([]*GroupMember, error) {
+	var gm []*GroupMember
+	err := json.Unmarshal(input, &gm)
+	if err != nil {
+		return nil, err
+	}
+
+	return gm, nil
+}
+
+func getMembershipList(g []*GroupMember) string {
+	out := []string{}
+	for _, member := range g {
+		out = append(out, member.GUID)
+	}
+
+	return strings.Join(out, ",")
+}
+
+func (g *GroupMembership) getGroupMembers(client *winrm.Client) ([]*GroupMember, error) {
+	cmd := fmt.Sprintf("Get-ADGroupMember -Identity %q", g.GroupGUID)
+
+	result, err := RunWinRMCommand(client, []string{cmd}, true, true)
+	if err != nil {
+		return nil, fmt.Errorf("while running Get-ADGroupMember: %s", err)
+	} else if result.ExitCode != 0 {
+		return nil, fmt.Errorf("command Get-ADGroupMember exited with a non-zero exit code(%d), stderr: %s, stdout: %s", result.ExitCode, result.StdErr, result.Stdout)
+	}
+
+	if strings.TrimSpace(result.Stdout) == "" {
+		return []*GroupMember{}, nil
+	}
+
+	gm, err := unmarshalGroupMembership([]byte(result.Stdout))
+	if err != nil {
+		return nil, fmt.Errorf("while unmarshalling group membership response: %s", err)
+	}
+
+	return gm, nil
+}
+
+func (g *GroupMembership) bulkGroupMembersOp(client *winrm.Client, operation string, members []*GroupMember) error {
+	memberList := getMembershipList(members)
+	cmd := fmt.Sprintf("%s -Identity %q %s -Confirm:$false", operation, g.GroupGUID, memberList)
+
+	result, err := RunWinRMCommand(client, []string{cmd}, false, false)
+	if err != nil {
+		return fmt.Errorf("while running %s: %s", operation, err)
+	} else if result.ExitCode != 0 {
+		return fmt.Errorf("command %s exited with a non-zero exit code(%d), stderr: %s, stdout: %s", operation, result.ExitCode, result.StdErr, result.Stdout)
+	}
+
+	return nil
+}
+
+func (g *GroupMembership) addGroupMembers(client *winrm.Client, members []*GroupMember) error {
+	return g.bulkGroupMembersOp(client, "Add-ADGroupMember", members)
```

Comment:
Is a zero-length check for `members` needed here?

Reply:
You're right, I'll add this before I merge!